File: py_src/problem011.py
# Project Euler Problem 11 Solution
#
# Problem statement:
# In the 20×20 grid below, four numbers along a diagonal line
# have been marked in red (see `input_files/problem011.txt').
# The product of these numbers is 26 × 63 × 78 × 14 = 1788696.
# What is the greatest product of four adjacent numbers in the
# same direction (up, down, left, right, or diagonally) in the
# 20×20 grid?
#
# Solution description:
# Brute-force.
#
# Author: Daniel Schuette
# Date: 2019/02/24
# License: MIT (see ../LICENSE.md)
import time


def parse_matrix(path):
    """
    Doc string.
    """
    matrix = []

    with open(path) as f:
        for line in f:
            data = line.rstrip("\n").split(sep=" ")
            matrix.append(data)

    return matrix


def greatest_product(matrix):
    """
    Doc string.
    """
    nums = []  # temp stores data
    max_product = 0  # temp stores maximum product
    max_adjacent = 4  # number of adjacent ints to multiply

    # left/right
    for arr in matrix:
        for i in range(len(arr)-max_adjacent+1):
            prod = mult_over_list(arr[i:(i+max_adjacent)])
            if prod > max_product:
                max_product = prod

    # up/down
    for i in range(len(matrix)-max_adjacent+1):  # row
        for j in range(len(matrix[0])):  # col
            for k in range(i, (i+max_adjacent)):
                nums.append(matrix[k][j])
            prod = mult_over_list(nums)
            if prod > max_product:
                max_product = prod
            nums = []  # reset for next iteration

    # diagonal - down and left to right
    for i in range(len(matrix)-max_adjacent+1):  # row
        for j in range(len(matrix[0])-max_adjacent+1):  # col
            for idx, k in enumerate(range(i, (i+max_adjacent))):
                nums.append(matrix[k][j+idx])
            prod = mult_over_list(nums)
            if prod > max_product:
                max_product = prod
            nums = []  # reset for next iteration

    # diagonal - down and right to left
    for i in range(len(matrix)-max_adjacent+1):  # row
        for j in range(len(matrix[0])-1, -1, -1):  # col right to left!
            for idx, k in enumerate(range(i, (i+max_adjacent))):
                nums.append(matrix[k][j-idx])
            prod = mult_over_list(nums)
            if prod > max_product:
                max_product = prod
            nums = []  # reset for next iteration

    # return result
    return max_product


def mult_over_list(l):
    """
    Doc string.
    """
    product = 1
    for e in l:
        product *= int(e)
    return product


if __name__ == "__main__":
    # calculate solution and time it
    start = time.time()
    data = parse_matrix("../input_files/problem011.txt")
    solution = greatest_product(data)
    end = time.time()

    # print out results
    print(f"Solution: {solution}")
    print(f"Elapsed time: {(end - start):.6f}s")

File: rust_src/src/lib/problem011.rs
/* Project Euler Problem 11 Solution
 *
 * Problem statement:
 * In the 20×20 grid below, four numbers along a diagonal line
 * have been marked in red (see `inputs/problem011.in').
 * The product of these numbers is 26 × 63 × 78 × 14 = 1788696.
 * What is the greatest product of four adjacent numbers in the
 * same direction (up, down, left, right, or diagonally) in the
 * 20×20 grid?
 *
 * Solution description:
 * Brute-force.
 *
 * Author: Daniel Schuette
 * Date: 2019/03/03
 * License: MIT (see ../LICENSE.md)
 */
use std::fs;

pub fn parse_matrix(path: &str) -> Vec<Vec<i64>> {
    let data_str: String;
    let mut data: Vec<Vec<i64>> = vec![];

    // read data from file
    let file = fs::read_to_string(path);
    if file.is_ok() {
        data_str = file.unwrap();
    } else {
        panic!("err: {} does not exist", path);
    }

    // read data into matrix
    for line in data_str.split("\n") {
        let mut v: Vec<i64> = vec![];
        for num in line.split(" ") {
            let p = num.parse::<i64>();
            let d: i64;
            if p.is_ok() {
                d = p.unwrap();
                v.push(d);
            } else {
                return data; /* reached end of data str */
            }
        }
        data.push(v);
    }
    return data;
}

pub fn greatest_product(matrix: Vec<Vec<i64>>) -> i64 {
    let mut product: i64 = 0;
    let adj_digits = 4; /* number of adjacent digits to multiply */

    // calculate product from left to right
    for r_idx in 0..(matrix.len()) {
        for c_idx in 0..(matrix[0].len() - adj_digits) {
            let p = mult_over_slice(&(matrix[r_idx][c_idx..(c_idx + adj_digits)]));
            if p > product {
                product = p;
            }
        }
    }

    // calculate product from top to bottom
    for r_idx in 0..(matrix.len() - adj_digits) {
        for c_idx in 0..(matrix[0].len()) {
            let mut v = vec![];
            for i in 0..adj_digits {
                v.push(matrix[r_idx + i][c_idx]);
            }

            let p = mult_over_slice(&v[..]);
            if p > product {
                product = p;
            }
        }
    }

    // calculate product from top left to bottom right
    for r_idx in 0..(matrix.len() - adj_digits) {
        for c_idx in 0..(matrix[0].len() - adj_digits) {
            let mut v = vec![];
            for i in 0..adj_digits {
                v.push(matrix[r_idx + i][c_idx + i]);
            }

            let p = mult_over_slice(&v[..]);
            if p > product {
                product = p;
            }
        }
    }

    // calculate product from top right to bottom left
    for r_idx in 0..(matrix.len() - adj_digits) {
        for c_idx in ((adj_digits - 1)..matrix[0].len()).rev() {
            let mut v = vec![];
            for i in 0..adj_digits {
                v.push(matrix[r_idx + i][c_idx - i]);
            }

            let p = mult_over_slice(&v[..]);
            if p > product {
                product = p;
            }
        }
    }

    return product;
}

pub fn mult_over_slice(s: &[i64]) -> i64 {
    let mut p: i64 = 1;

    for i in s {
        p *= i;
    }
    return p;
}
